### ISSUE TITLE
IDAS performance regression

### ISSUE THREAD

**Issue by jgillis** [casadi/casadi#3139]
Used to work, now yields `.../casadi/interfaces/sundials/idas_interface.cpp:605: IDASolveB returned "IDA_TOO_MUCH_WORK". Consult IDAS documentation.`

```python

from casadi import *

x1 = MX.sym("x1")
x2 = MX.sym("x2")
x = vertcat(x1,x2)

z = MX.sym("z")

u = MX.sym("u")
DT = MX.sym("DT")


dae = {"x": x,"ode":DT*vertcat(z * x1 - x2 + u,x1), "p": vertcat(u,DT), "z": z, "alg": z-(1 - x2**2), "quad": DT*(x1**2 + x2**2 + u**2)}

intg = integrator("intg","idas",dae,{"print_in":True,"expand":True})

X0 = vertcat(MX.sym("X"),MX.sym("X"))
U0 = MX.sym("U0")

X1_intg = intg(x0=X0,p=vertcat(U0,10))

f = Function("f",[vertcat(X0,U0),MX(0,1)],[X1_intg["qf"],0],["x","p"],["f","g"])

hess = f.factory("ff",["x", "p", "lam:f", "lam:g"], ["transpose:triu:hess:gamma:x:x"],{"gamma":["f","g"]})


args = {}
args["x"] = vertcat(0,1,1)
args["lam_f"] = 1

print(hess(**args))
```

**Comment by jgillis** [comment]
Reduced
```python
from casadi import *

x1 = MX.sym("x1")
x2 = MX.sym("x2")
x = vertcat(x1,x2)

z = MX.sym("z")

u = MX.sym("u")
DT = MX.sym("DT")


dae = {"x": x,"ode":DT*vertcat(z * x1 - x2 + u,x1), "p": vertcat(u,DT), "z": z, "alg": z-(1 - x2**2), "quad": DT*(x1**2 + x2**2 + u**2)}

intg = integrator("intg","idas",dae)

X0 = vertcat(MX.sym("X"),MX.sym("X"))
U0 = MX.sym("U0")

X1_intg = intg(x0=X0,p=vertcat(U0,10))

f = Function("f",[vertcat(X0,U0)],[X1_intg["qf"]],["x"],["f"])

h=f.reverse(1).forward(1)

print(h)
print(h(vertcat(0,1,1),0,1,0,vertcat(1,0,0),0,1))
```

**Comment by jaeandersson** [comment]
Interesting. I'll do another git-bisect

**Comment by jaeandersson** [comment]
Weird...

8a3a7d5733189f0a034d023c9559aee736cf77a4 is the first bad commit
commit 8a3a7d5733189f0a034d023c9559aee736cf77a4
Author: Joel Andersson <joel@jaeandersson.com>
Date:   Fri Apr 21 16:07:00 2023 -0500

    Issue #3121 Fixed IDAS bug causing performance decay
    Bug was introduced in af60b3c

 casadi/interfaces/sundials/idas_interface.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

**Comment by jaeandersson** [comment]
The bug appeared to be present in 53004904772abbcdebeaba34112c2e16f07e6891, before the bug that was fixed in 8a3a7d5733189f0a034d023c9559aee736cf77a4 was introduced. The af60b3c bug seems to have hidden the bug and then it resurfaced when the bug was fixed

Should do another git bisect between CasADi 3.5.5 and 53004904772abbcdebeaba34112c2e16f07e6891

**Comment by jaeandersson** [comment]
Looks like the change happens in a simplification step. The directional derivative of (dg/dx * fwd_x) with respect to fwd_x is replaced with a directional derivative of g(x).
